File: backend/app/schemas/medicine_pricing.py
"""
Medicine Pricing schemas — Pydantic validation for the pricing module.
"""

from pydantic import BaseModel, Field, model_validator
from datetime import date, datetime
from typing import Optional, List
from enum import Enum


class PricingMode(str, Enum):
    PCT_MARGE = "pct_marge"
    MANUEL = "manuel"
    CARTON_FIXE = "carton_fixe"


class OrdonnanceType(str, Enum):
    NON = "non"
    OUI = "oui"
    LISTE1 = "liste1"
    LISTE2 = "liste2"
    STUP = "stup"


class MedicinePricingCreate(BaseModel):
    """Schema for creating a medicine pricing entry."""
    nom: str = Field(..., min_length=1, max_length=200, description="Nom du médicament")
    dci: Optional[str] = Field(None, max_length=200, description="DCI")
    forme: Optional[str] = Field(None, max_length=100, description="Forme galénique")
    dosage: Optional[str] = Field(None, max_length=100, description="Dosage")
    lot: str = Field(..., min_length=1, max_length=100, description="Numéro de lot")
    fournisseur: Optional[str] = Field(None, max_length=200)
    bon_livraison: Optional[str] = Field(None, max_length=100)
    date_reception: Optional[date] = None
    date_peremption: Optional[date] = None

    # Conditionnement
    nb_cartons: int = Field(..., gt=0, description="Nombre de cartons reçus")
    boites_par_carton: int = Field(..., gt=0, description="Boîtes par carton")
    plaquettes_par_boite: int = Field(..., gt=0, description="Plaquettes par boîte")
    comprimes_par_plaquette: int = Field(..., gt=0, description="Comprimés par plaquette")

    # Prix
    prix_mode: PricingMode = Field(default=PricingMode.MANUEL)
    achat_carton: float = Field(..., gt=0, description="Prix d'achat du carton (FBu)")
    achat_boite: float = Field(default=0.0, ge=0, description="Prix d'achat par boîte")
    achat_plaquette: float = Field(default=0.0, ge=0, description="Prix d'achat par plaquette")
    achat_comprime: float = Field(default=0.0, ge=0, description="Prix d'achat par unité")
    vente_carton: float = Field(default=0.0, ge=0)
    vente_boite: float = Field(default=0.0, ge=0)
    vente_plaquette: float = Field(default=0.0, ge=0)
    vente_comprime: float = Field(default=0.0, ge=0)
    marge_pct: Optional[float] = Field(default=None, ge=0)

    # Stock & Alertes
    seuil_alerte: int = Field(default=10, ge=0)
    seuil_niveau: str = Field(default="comprimes", description="Niveau du seuil: comprimes, plaquettes, boites, cartons")
    emplacement: Optional[str] = Field(None, max_length=100)
    alerte_peremption: bool = Field(default=True)
    alerte_jours: Optional[int] = Field(default=30, ge=1, description="Nombre de jours avant expiration pour l'alerte")
    ordonnance: OrdonnanceType = Field(default=OrdonnanceType.NON)

    @model_validator(mode='after')
    def validate_dates(self):
        if self.date_reception and self.date_peremption:
            if self.date_peremption <= self.date_reception:
                raise ValueError("La date de péremption doit être postérieure à la date de réception")
        return self

    @model_validator(mode='after')
    def validate_expiry_future(self):
        """R6: La date d'expiration doit être dans le futur."""
        if self.date_peremption and self.date_peremption <= date.today():
            raise ValueError("La date de péremption doit être dans le futur")
        return self

    @model_validator(mode='after')
    def validate_pricing_mode(self):
        if self.prix_mode == PricingMode.PCT_MARGE:
            if self.marge_pct is None or self.marge_pct <= 0:
                raise ValueError("Le pourcentage de marge est requis pour le mode 'Pourcentage de marge'")
        elif self.prix_mode == PricingMode.MANUEL:
            if self.vente_carton <= 0 or self.vente_boite <= 0 or self.vente_plaquette <= 0 or self.vente_comprime <= 0:
                raise ValueError("Tous les prix de vente sont requis pour le mode 'Manuel'")
        elif self.prix_mode == PricingMode.CARTON_FIXE:
            if self.vente_carton <= 0:
                raise ValueError("Le prix de vente du carton est requis pour le mode 'Carton fixé'")
        return self

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "nom": "Amoxicilline 500mg",
                    "lot": "LOT-2026-001",
                    "nb_cartons": 10,
                    "boites_par_carton": 20,
                    "plaquettes_par_boite": 3,
                    "comprimes_par_plaquette": 8,
                    "prix_mode": "pct_marge",
                    "achat_carton": 120000,
                    "marge_pct": 25,
                }
            ]
        }
    }


class MedicinePricingUpdate(BaseModel):
    """Schema for updating a pricing entry (all fields optional)."""
    nom: Optional[str] = Field(None, min_length=1, max_length=200)
    dci: Optional[str] = None
    forme: Optional[str] = None
    dosage: Optional[str] = None
    lot: Optional[str] = Field(None, min_length=1, max_length=100)
    fournisseur: Optional[str] = None
    bon_livraison: Optional[str] = None
    date_reception: Optional[date] = None
    date_peremption: Optional[date] = None
    nb_cartons: Optional[int] = Field(None, gt=0)
    boites_par_carton: Optional[int] = Field(None, gt=0)
    plaquettes_par_boite: Optional[int] = Field(None, gt=0)
    comprimes_par_plaquette: Optional[int] = Field(None, gt=0)
    prix_mode: Optional[PricingMode] = None
    achat_carton: Optional[float] = Field(None, gt=0)
    achat_boite: Optional[float] = Field(None, ge=0)
    achat_plaquette: Optional[float] = Field(None, ge=0)
    achat_comprime: Optional[float] = Field(None, ge=0)
    vente_carton: Optional[float] = Field(None, ge=0)
    vente_boite: Optional[float] = Field(None, ge=0)
    vente_plaquette: Optional[float] = Field(None, ge=0)
    vente_comprime: Optional[float] = Field(None, ge=0)
    marge_pct: Optional[float] = Field(None, ge=0)
    seuil_alerte: Optional[int] = Field(None, ge=0)
    seuil_niveau: Optional[str] = None
    emplacement: Optional[str] = None
    alerte_peremption: Optional[bool] = None
    alerte_jours: Optional[int] = Field(None, ge=1)
    ordonnance: Optional[OrdonnanceType] = None


class MedicinePricingResponse(BaseModel):
    """Schema for pricing entry in responses."""
    id: int
    medicine_id: Optional[int] = None
    nom: str
    dci: Optional[str]
    forme: Optional[str]
    dosage: Optional[str]
    lot: str
    fournisseur: Optional[str]
    bon_livraison: Optional[str]
    date_reception: Optional[date]
    date_peremption: Optional[date]

    nb_cartons: int
    boites_par_carton: int
    plaquettes_par_boite: int
    comprimes_par_plaquette: int
    total_boites: int
    total_plaquettes: int
    total_comprimes: int

    prix_mode: str
    achat_carton: float
    achat_boite: float = 0.0
    achat_plaquette: float = 0.0
    achat_comprime: float = 0.0
    vente_carton: float
    vente_boite: float
    vente_plaquette: float
    vente_comprime: float
    marge_pct: Optional[float]
    benefice_estime: float

    seuil_alerte: int
    seuil_niveau: str = "comprimes"
    emplacement: Optional[str]
    alerte_peremption: bool
    alerte_jours: Optional[int] = 30
    ordonnance: str

    created_at: datetime
    updated_at: datetime

    # Calculated alert fields
    expire_bientot: bool = False
    stock_faible: bool = False

    model_config = {"from_attributes": True}

File: frontend1/lib/models/medicine_pricing.dart
/// Modes de fixation des prix
enum PricingMode {
  pctMarge('pct_marge', 'Pourcentage de marge'),
  manuel('manuel', 'Saisie manuelle'),
  cartonFixe('carton_fixe', 'Prix carton fixé');

  final String value;
  final String label;
  const PricingMode(this.value, this.label);

  static PricingMode fromString(String value) {
    return PricingMode.values.firstWhere(
      (e) => e.value == value,
      orElse: () => PricingMode.pctMarge,
    );
  }
}

/// Types d'ordonnance
enum OrdonnanceType {
  non('non', 'Sans ordonnance'),
  oui('oui', 'Avec ordonnance'),
  liste1('liste1', 'Liste I (toxiques)'),
  liste2('liste2', 'Liste II (dangereux)'),
  stup('stup', 'Stupéfiants');

  final String value;
  final String label;
  const OrdonnanceType(this.value, this.label);

  static OrdonnanceType fromString(String value) {
    return OrdonnanceType.values.firstWhere(
      (e) => e.value == value,
      orElse: () => OrdonnanceType.non,
    );
  }
}

/// Modèle de données pour une entrée de prix médicament
class MedicinePricing {
  final int id;
  final int? medicineId;
  final String nom;
  final String? dci;
  final String? forme;
  final String? dosage;
  final String lot;
  final String? fournisseur;
  final String? bonLivraison;
  final DateTime? dateReception;
  final DateTime? datePeremption;

  // Conditionnement
  final int nbCartons;
  final int boitesParCarton;
  final int plaquettesParBoite;
  final int comprimesParPlaquette;
  final int totalBoites;
  final int totalPlaquettes;
  final int totalComprimes;

  // Prix
  final String prixMode;
  final double achatCarton;
  final double achatBoite;
  final double achatPlaquette;
  final double achatComprime;
  final double venteCarton;
  final double venteBoite;
  final double ventePlaquette;
  final double venteComprime;
  final double? margePct;
  final double beneficeEstime;

  // Stock & Alertes
  final int seuilAlerte;
  final String seuilNiveau;
  final String? emplacement;
  final bool alertePeremption;
  final int? alerteJours;
  final String ordonnance;

  final DateTime createdAt;
  final DateTime updatedAt;

  // Calculated alerts
  final bool expireBientot;
  final bool stockFaible;

  MedicinePricing({
    required this.id,
    this.medicineId,
    required this.nom,
    this.dci,
    this.forme,
    this.dosage,
    required this.lot,
    this.fournisseur,
    this.bonLivraison,
    this.dateReception,
    this.datePeremption,
    required this.nbCartons,
    required this.boitesParCarton,
    required this.plaquettesParBoite,
    required this.comprimesParPlaquette,
    required this.totalBoites,
    required this.totalPlaquettes,
    required this.totalComprimes,
    required this.prixMode,
    required this.achatCarton,
    this.achatBoite = 0.0,
    this.achatPlaquette = 0.0,
    this.achatComprime = 0.0,
    required this.venteCarton,
    required this.venteBoite,
    required this.ventePlaquette,
    required this.venteComprime,
    this.margePct,
    required this.beneficeEstime,
    required this.seuilAlerte,
    this.seuilNiveau = 'comprimes',
    this.emplacement,
    required this.alertePeremption,
    this.alerteJours,
    required this.ordonnance,
    required this.createdAt,
    required this.updatedAt,
    this.expireBientot = false,
    this.stockFaible = false,
  });

  factory MedicinePricing.fromJson(Map<String, dynamic> json) {
    return MedicinePricing(
      id: json['id'] as int,
      medicineId: json['medicine_id'] as int?,
      nom: json['nom'] as String,
      dci: json['dci'] as String?,
      forme: json['forme'] as String?,
      dosage: json['dosage'] as String?,
      lot: json['lot'] as String,
      fournisseur: json['fournisseur'] as String?,
      bonLivraison: json['bon_livraison'] as String?,
      dateReception: json['date_reception'] != null
          ? DateTime.parse(json['date_reception'] as String)
          : null,
      datePeremption: json['date_peremption'] != null
          ? DateTime.parse(json['date_peremption'] as String)
          : null,
      nbCartons: json['nb_cartons'] as int,
      boitesParCarton: json['boites_par_carton'] as int,
      plaquettesParBoite: json['plaquettes_par_boite'] as int,
      comprimesParPlaquette: json['comprimes_par_plaquette'] as int,
      totalBoites: json['total_boites'] as int,
      totalPlaquettes: json['total_plaquettes'] as int,
      totalComprimes: json['total_comprimes'] as int,
      prixMode: json['prix_mode'] as String,
      achatCarton: (json['achat_carton'] as num).toDouble(),
      achatBoite: (json['achat_boite'] as num?)?.toDouble() ?? 0.0,
      achatPlaquette: (json['achat_plaquette'] as num?)?.toDouble() ?? 0.0,
      achatComprime: (json['achat_comprime'] as num?)?.toDouble() ?? 0.0,
      venteCarton: (json['vente_carton'] as num).toDouble(),
      venteBoite: (json['vente_boite'] as num).toDouble(),
      ventePlaquette: (json['vente_plaquette'] as num).toDouble(),
      venteComprime: (json['vente_comprime'] as num).toDouble(),
      margePct: json['marge_pct'] != null
          ? (json['marge_pct'] as num).toDouble()
          : null,
      beneficeEstime: (json['benefice_estime'] as num).toDouble(),
      seuilAlerte: json['seuil_alerte'] as int,
      seuilNiveau: json['seuil_niveau'] as String? ?? 'comprimes',
      emplacement: json['emplacement'] as String?,
      alertePeremption: json['alerte_peremption'] as bool,
      alerteJours: json['alerte_jours'] as int?,
      ordonnance: json['ordonnance'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      expireBientot: json['expire_bientot'] as bool? ?? false,
      stockFaible: json['stock_faible'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'nom': nom,
      'dci': dci,
      'forme': forme,
      'dosage': dosage,
      'lot': lot,
      'fournisseur': fournisseur,
      'bon_livraison': bonLivraison,
      'date_reception': dateReception?.toIso8601String().split('T').first,
      'date_peremption': datePeremption?.toIso8601String().split('T').first,
      'nb_cartons': nbCartons,
      'boites_par_carton': boitesParCarton,
      'plaquettes_par_boite': plaquettesParBoite,
      'comprimes_par_plaquette': comprimesParPlaquette,
      'prix_mode': prixMode,
      'achat_carton': achatCarton,
      'achat_boite': achatBoite,
      'achat_plaquette': achatPlaquette,
      'achat_comprime': achatComprime,
      'vente_carton': venteCarton,
      'vente_boite': venteBoite,
      'vente_plaquette': ventePlaquette,
      'vente_comprime': venteComprime,
      'marge_pct': margePct,
      'seuil_alerte': seuilAlerte,
      'seuil_niveau': seuilNiveau,
      'emplacement': emplacement,
      'alerte_peremption': alertePeremption,
      'alerte_jours': alerteJours,
      'ordonnance': ordonnance,
    };
  }
}

/// Réponse paginée des entrées de prix
class PricingResponse {
  final List<MedicinePricing> items;
  final int total;
  final int totalPages;

  PricingResponse({
    required this.items,
    required this.total,
    required this.totalPages,
  });

  factory PricingResponse.fromJson(Map<String, dynamic> json) {
    return PricingResponse(
      items: (json['items'] as List)
          .map((item) => MedicinePricing.fromJson(item as Map<String, dynamic>))
          .toList(),
      total: json['total'] as int,
      totalPages: json['total_pages'] as int,
    );
  }
}
